--- external/F746_Gameboy/inc/z80_opcode_goto.h	2020-10-13 23:22:04.546294215 +0200
+++ external/F746_Gameboy_git/inc/z80_opcode_goto.h	2020-10-13 21:42:07.267592392 +0200
@@ -0,0 +1,2600 @@
+//! This file is auto generated, don't edit
+//! Generated on: 2020-04-24 19:21:50.493433
+
+#ifndef Z80_OPCODE_GOTO_H_
+#define Z80_OPCODE_GOTO_H_
+
+#if OPCODE_GOTO
+
+#include "z80_ub.h"
+#include "z80_opcode_func.h"
+#include "z80_opcode_cb_func.h"
+
+static inline void z80_single_step(void)
+{
+  #define EXECUTE()            goto *opcode0_labels[z80.opcode];
+  #define EXECUTE_CB()         goto *opcode1_labels[z80.opcode];
+  #define EXECUTE_END()        goto single_step_end;
+  #define EXECUTE_CB_END()     goto single_step_cb_end;
+
+  static const void *const opcode0_labels[] = {
+    &&op0_entry_0x00,
+    &&op0_entry_0x01,
+    &&op0_entry_0x02,
+    &&op0_entry_0x03,
+    &&op0_entry_0x04,
+    &&op0_entry_0x05,
+    &&op0_entry_0x06,
+    &&op0_entry_0x07,
+    &&op0_entry_0x08,
+    &&op0_entry_0x09,
+    &&op0_entry_0x0a,
+    &&op0_entry_0x0b,
+    &&op0_entry_0x0c,
+    &&op0_entry_0x0d,
+    &&op0_entry_0x0e,
+    &&op0_entry_0x0f,
+    &&op0_entry_0x10,
+    &&op0_entry_0x11,
+    &&op0_entry_0x12,
+    &&op0_entry_0x13,
+    &&op0_entry_0x14,
+    &&op0_entry_0x15,
+    &&op0_entry_0x16,
+    &&op0_entry_0x17,
+    &&op0_entry_0x18,
+    &&op0_entry_0x19,
+    &&op0_entry_0x1a,
+    &&op0_entry_0x1b,
+    &&op0_entry_0x1c,
+    &&op0_entry_0x1d,
+    &&op0_entry_0x1e,
+    &&op0_entry_0x1f,
+    &&op0_entry_0x20,
+    &&op0_entry_0x21,
+    &&op0_entry_0x22,
+    &&op0_entry_0x23,
+    &&op0_entry_0x24,
+    &&op0_entry_0x25,
+    &&op0_entry_0x26,
+    &&op0_entry_0x27,
+    &&op0_entry_0x28,
+    &&op0_entry_0x29,
+    &&op0_entry_0x2a,
+    &&op0_entry_0x2b,
+    &&op0_entry_0x2c,
+    &&op0_entry_0x2d,
+    &&op0_entry_0x2e,
+    &&op0_entry_0x2f,
+    &&op0_entry_0x30,
+    &&op0_entry_0x31,
+    &&op0_entry_0x32,
+    &&op0_entry_0x33,
+    &&op0_entry_0x34,
+    &&op0_entry_0x35,
+    &&op0_entry_0x36,
+    &&op0_entry_0x37,
+    &&op0_entry_0x38,
+    &&op0_entry_0x39,
+    &&op0_entry_0x3a,
+    &&op0_entry_0x3b,
+    &&op0_entry_0x3c,
+    &&op0_entry_0x3d,
+    &&op0_entry_0x3e,
+    &&op0_entry_0x3f,
+    &&op0_entry_0x40,
+    &&op0_entry_0x41,
+    &&op0_entry_0x42,
+    &&op0_entry_0x43,
+    &&op0_entry_0x44,
+    &&op0_entry_0x45,
+    &&op0_entry_0x46,
+    &&op0_entry_0x47,
+    &&op0_entry_0x48,
+    &&op0_entry_0x49,
+    &&op0_entry_0x4a,
+    &&op0_entry_0x4b,
+    &&op0_entry_0x4c,
+    &&op0_entry_0x4d,
+    &&op0_entry_0x4e,
+    &&op0_entry_0x4f,
+    &&op0_entry_0x50,
+    &&op0_entry_0x51,
+    &&op0_entry_0x52,
+    &&op0_entry_0x53,
+    &&op0_entry_0x54,
+    &&op0_entry_0x55,
+    &&op0_entry_0x56,
+    &&op0_entry_0x57,
+    &&op0_entry_0x58,
+    &&op0_entry_0x59,
+    &&op0_entry_0x5a,
+    &&op0_entry_0x5b,
+    &&op0_entry_0x5c,
+    &&op0_entry_0x5d,
+    &&op0_entry_0x5e,
+    &&op0_entry_0x5f,
+    &&op0_entry_0x60,
+    &&op0_entry_0x61,
+    &&op0_entry_0x62,
+    &&op0_entry_0x63,
+    &&op0_entry_0x64,
+    &&op0_entry_0x65,
+    &&op0_entry_0x66,
+    &&op0_entry_0x67,
+    &&op0_entry_0x68,
+    &&op0_entry_0x69,
+    &&op0_entry_0x6a,
+    &&op0_entry_0x6b,
+    &&op0_entry_0x6c,
+    &&op0_entry_0x6d,
+    &&op0_entry_0x6e,
+    &&op0_entry_0x6f,
+    &&op0_entry_0x70,
+    &&op0_entry_0x71,
+    &&op0_entry_0x72,
+    &&op0_entry_0x73,
+    &&op0_entry_0x74,
+    &&op0_entry_0x75,
+    &&op0_entry_0x76,
+    &&op0_entry_0x77,
+    &&op0_entry_0x78,
+    &&op0_entry_0x79,
+    &&op0_entry_0x7a,
+    &&op0_entry_0x7b,
+    &&op0_entry_0x7c,
+    &&op0_entry_0x7d,
+    &&op0_entry_0x7e,
+    &&op0_entry_0x7f,
+    &&op0_entry_0x80,
+    &&op0_entry_0x81,
+    &&op0_entry_0x82,
+    &&op0_entry_0x83,
+    &&op0_entry_0x84,
+    &&op0_entry_0x85,
+    &&op0_entry_0x86,
+    &&op0_entry_0x87,
+    &&op0_entry_0x88,
+    &&op0_entry_0x89,
+    &&op0_entry_0x8a,
+    &&op0_entry_0x8b,
+    &&op0_entry_0x8c,
+    &&op0_entry_0x8d,
+    &&op0_entry_0x8e,
+    &&op0_entry_0x8f,
+    &&op0_entry_0x90,
+    &&op0_entry_0x91,
+    &&op0_entry_0x92,
+    &&op0_entry_0x93,
+    &&op0_entry_0x94,
+    &&op0_entry_0x95,
+    &&op0_entry_0x96,
+    &&op0_entry_0x97,
+    &&op0_entry_0x98,
+    &&op0_entry_0x99,
+    &&op0_entry_0x9a,
+    &&op0_entry_0x9b,
+    &&op0_entry_0x9c,
+    &&op0_entry_0x9d,
+    &&op0_entry_0x9e,
+    &&op0_entry_0x9f,
+    &&op0_entry_0xa0,
+    &&op0_entry_0xa1,
+    &&op0_entry_0xa2,
+    &&op0_entry_0xa3,
+    &&op0_entry_0xa4,
+    &&op0_entry_0xa5,
+    &&op0_entry_0xa6,
+    &&op0_entry_0xa7,
+    &&op0_entry_0xa8,
+    &&op0_entry_0xa9,
+    &&op0_entry_0xaa,
+    &&op0_entry_0xab,
+    &&op0_entry_0xac,
+    &&op0_entry_0xad,
+    &&op0_entry_0xae,
+    &&op0_entry_0xaf,
+    &&op0_entry_0xb0,
+    &&op0_entry_0xb1,
+    &&op0_entry_0xb2,
+    &&op0_entry_0xb3,
+    &&op0_entry_0xb4,
+    &&op0_entry_0xb5,
+    &&op0_entry_0xb6,
+    &&op0_entry_0xb7,
+    &&op0_entry_0xb8,
+    &&op0_entry_0xb9,
+    &&op0_entry_0xba,
+    &&op0_entry_0xbb,
+    &&op0_entry_0xbc,
+    &&op0_entry_0xbd,
+    &&op0_entry_0xbe,
+    &&op0_entry_0xbf,
+    &&op0_entry_0xc0,
+    &&op0_entry_0xc1,
+    &&op0_entry_0xc2,
+    &&op0_entry_0xc3,
+    &&op0_entry_0xc4,
+    &&op0_entry_0xc5,
+    &&op0_entry_0xc6,
+    &&op0_entry_0xc7,
+    &&op0_entry_0xc8,
+    &&op0_entry_0xc9,
+    &&op0_entry_0xca,
+    &&op0_entry_0xcb,
+    &&op0_entry_0xcc,
+    &&op0_entry_0xcd,
+    &&op0_entry_0xce,
+    &&op0_entry_0xcf,
+    &&op0_entry_0xd0,
+    &&op0_entry_0xd1,
+    &&op0_entry_0xd2,
+    &&op0_entry_0xd3,
+    &&op0_entry_0xd4,
+    &&op0_entry_0xd5,
+    &&op0_entry_0xd6,
+    &&op0_entry_0xd7,
+    &&op0_entry_0xd8,
+    &&op0_entry_0xd9,
+    &&op0_entry_0xda,
+    &&op0_entry_0xdb,
+    &&op0_entry_0xdc,
+    &&op0_entry_0xdd,
+    &&op0_entry_0xde,
+    &&op0_entry_0xdf,
+    &&op0_entry_0xe0,
+    &&op0_entry_0xe1,
+    &&op0_entry_0xe2,
+    &&op0_entry_0xe3,
+    &&op0_entry_0xe4,
+    &&op0_entry_0xe5,
+    &&op0_entry_0xe6,
+    &&op0_entry_0xe7,
+    &&op0_entry_0xe8,
+    &&op0_entry_0xe9,
+    &&op0_entry_0xea,
+    &&op0_entry_0xeb,
+    &&op0_entry_0xec,
+    &&op0_entry_0xed,
+    &&op0_entry_0xee,
+    &&op0_entry_0xef,
+    &&op0_entry_0xf0,
+    &&op0_entry_0xf1,
+    &&op0_entry_0xf2,
+    &&op0_entry_0xf3,
+    &&op0_entry_0xf4,
+    &&op0_entry_0xf5,
+    &&op0_entry_0xf6,
+    &&op0_entry_0xf7,
+    &&op0_entry_0xf8,
+    &&op0_entry_0xf9,
+    &&op0_entry_0xfa,
+    &&op0_entry_0xfb,
+    &&op0_entry_0xfc,
+    &&op0_entry_0xfd,
+    &&op0_entry_0xfe,
+    &&op0_entry_0xff,
+  };
+
+  static const void *const opcode1_labels[] = {
+    &&op1_entry_0x00,
+    &&op1_entry_0x01,
+    &&op1_entry_0x02,
+    &&op1_entry_0x03,
+    &&op1_entry_0x04,
+    &&op1_entry_0x05,
+    &&op1_entry_0x06,
+    &&op1_entry_0x07,
+    &&op1_entry_0x08,
+    &&op1_entry_0x09,
+    &&op1_entry_0x0a,
+    &&op1_entry_0x0b,
+    &&op1_entry_0x0c,
+    &&op1_entry_0x0d,
+    &&op1_entry_0x0e,
+    &&op1_entry_0x0f,
+    &&op1_entry_0x10,
+    &&op1_entry_0x11,
+    &&op1_entry_0x12,
+    &&op1_entry_0x13,
+    &&op1_entry_0x14,
+    &&op1_entry_0x15,
+    &&op1_entry_0x16,
+    &&op1_entry_0x17,
+    &&op1_entry_0x18,
+    &&op1_entry_0x19,
+    &&op1_entry_0x1a,
+    &&op1_entry_0x1b,
+    &&op1_entry_0x1c,
+    &&op1_entry_0x1d,
+    &&op1_entry_0x1e,
+    &&op1_entry_0x1f,
+    &&op1_entry_0x20,
+    &&op1_entry_0x21,
+    &&op1_entry_0x22,
+    &&op1_entry_0x23,
+    &&op1_entry_0x24,
+    &&op1_entry_0x25,
+    &&op1_entry_0x26,
+    &&op1_entry_0x27,
+    &&op1_entry_0x28,
+    &&op1_entry_0x29,
+    &&op1_entry_0x2a,
+    &&op1_entry_0x2b,
+    &&op1_entry_0x2c,
+    &&op1_entry_0x2d,
+    &&op1_entry_0x2e,
+    &&op1_entry_0x2f,
+    &&op1_entry_0x30,
+    &&op1_entry_0x31,
+    &&op1_entry_0x32,
+    &&op1_entry_0x33,
+    &&op1_entry_0x34,
+    &&op1_entry_0x35,
+    &&op1_entry_0x36,
+    &&op1_entry_0x37,
+    &&op1_entry_0x38,
+    &&op1_entry_0x39,
+    &&op1_entry_0x3a,
+    &&op1_entry_0x3b,
+    &&op1_entry_0x3c,
+    &&op1_entry_0x3d,
+    &&op1_entry_0x3e,
+    &&op1_entry_0x3f,
+    &&op1_entry_0x40,
+    &&op1_entry_0x41,
+    &&op1_entry_0x42,
+    &&op1_entry_0x43,
+    &&op1_entry_0x44,
+    &&op1_entry_0x45,
+    &&op1_entry_0x46,
+    &&op1_entry_0x47,
+    &&op1_entry_0x48,
+    &&op1_entry_0x49,
+    &&op1_entry_0x4a,
+    &&op1_entry_0x4b,
+    &&op1_entry_0x4c,
+    &&op1_entry_0x4d,
+    &&op1_entry_0x4e,
+    &&op1_entry_0x4f,
+    &&op1_entry_0x50,
+    &&op1_entry_0x51,
+    &&op1_entry_0x52,
+    &&op1_entry_0x53,
+    &&op1_entry_0x54,
+    &&op1_entry_0x55,
+    &&op1_entry_0x56,
+    &&op1_entry_0x57,
+    &&op1_entry_0x58,
+    &&op1_entry_0x59,
+    &&op1_entry_0x5a,
+    &&op1_entry_0x5b,
+    &&op1_entry_0x5c,
+    &&op1_entry_0x5d,
+    &&op1_entry_0x5e,
+    &&op1_entry_0x5f,
+    &&op1_entry_0x60,
+    &&op1_entry_0x61,
+    &&op1_entry_0x62,
+    &&op1_entry_0x63,
+    &&op1_entry_0x64,
+    &&op1_entry_0x65,
+    &&op1_entry_0x66,
+    &&op1_entry_0x67,
+    &&op1_entry_0x68,
+    &&op1_entry_0x69,
+    &&op1_entry_0x6a,
+    &&op1_entry_0x6b,
+    &&op1_entry_0x6c,
+    &&op1_entry_0x6d,
+    &&op1_entry_0x6e,
+    &&op1_entry_0x6f,
+    &&op1_entry_0x70,
+    &&op1_entry_0x71,
+    &&op1_entry_0x72,
+    &&op1_entry_0x73,
+    &&op1_entry_0x74,
+    &&op1_entry_0x75,
+    &&op1_entry_0x76,
+    &&op1_entry_0x77,
+    &&op1_entry_0x78,
+    &&op1_entry_0x79,
+    &&op1_entry_0x7a,
+    &&op1_entry_0x7b,
+    &&op1_entry_0x7c,
+    &&op1_entry_0x7d,
+    &&op1_entry_0x7e,
+    &&op1_entry_0x7f,
+    &&op1_entry_0x80,
+    &&op1_entry_0x81,
+    &&op1_entry_0x82,
+    &&op1_entry_0x83,
+    &&op1_entry_0x84,
+    &&op1_entry_0x85,
+    &&op1_entry_0x86,
+    &&op1_entry_0x87,
+    &&op1_entry_0x88,
+    &&op1_entry_0x89,
+    &&op1_entry_0x8a,
+    &&op1_entry_0x8b,
+    &&op1_entry_0x8c,
+    &&op1_entry_0x8d,
+    &&op1_entry_0x8e,
+    &&op1_entry_0x8f,
+    &&op1_entry_0x90,
+    &&op1_entry_0x91,
+    &&op1_entry_0x92,
+    &&op1_entry_0x93,
+    &&op1_entry_0x94,
+    &&op1_entry_0x95,
+    &&op1_entry_0x96,
+    &&op1_entry_0x97,
+    &&op1_entry_0x98,
+    &&op1_entry_0x99,
+    &&op1_entry_0x9a,
+    &&op1_entry_0x9b,
+    &&op1_entry_0x9c,
+    &&op1_entry_0x9d,
+    &&op1_entry_0x9e,
+    &&op1_entry_0x9f,
+    &&op1_entry_0xa0,
+    &&op1_entry_0xa1,
+    &&op1_entry_0xa2,
+    &&op1_entry_0xa3,
+    &&op1_entry_0xa4,
+    &&op1_entry_0xa5,
+    &&op1_entry_0xa6,
+    &&op1_entry_0xa7,
+    &&op1_entry_0xa8,
+    &&op1_entry_0xa9,
+    &&op1_entry_0xaa,
+    &&op1_entry_0xab,
+    &&op1_entry_0xac,
+    &&op1_entry_0xad,
+    &&op1_entry_0xae,
+    &&op1_entry_0xaf,
+    &&op1_entry_0xb0,
+    &&op1_entry_0xb1,
+    &&op1_entry_0xb2,
+    &&op1_entry_0xb3,
+    &&op1_entry_0xb4,
+    &&op1_entry_0xb5,
+    &&op1_entry_0xb6,
+    &&op1_entry_0xb7,
+    &&op1_entry_0xb8,
+    &&op1_entry_0xb9,
+    &&op1_entry_0xba,
+    &&op1_entry_0xbb,
+    &&op1_entry_0xbc,
+    &&op1_entry_0xbd,
+    &&op1_entry_0xbe,
+    &&op1_entry_0xbf,
+    &&op1_entry_0xc0,
+    &&op1_entry_0xc1,
+    &&op1_entry_0xc2,
+    &&op1_entry_0xc3,
+    &&op1_entry_0xc4,
+    &&op1_entry_0xc5,
+    &&op1_entry_0xc6,
+    &&op1_entry_0xc7,
+    &&op1_entry_0xc8,
+    &&op1_entry_0xc9,
+    &&op1_entry_0xca,
+    &&op1_entry_0xcb,
+    &&op1_entry_0xcc,
+    &&op1_entry_0xcd,
+    &&op1_entry_0xce,
+    &&op1_entry_0xcf,
+    &&op1_entry_0xd0,
+    &&op1_entry_0xd1,
+    &&op1_entry_0xd2,
+    &&op1_entry_0xd3,
+    &&op1_entry_0xd4,
+    &&op1_entry_0xd5,
+    &&op1_entry_0xd6,
+    &&op1_entry_0xd7,
+    &&op1_entry_0xd8,
+    &&op1_entry_0xd9,
+    &&op1_entry_0xda,
+    &&op1_entry_0xdb,
+    &&op1_entry_0xdc,
+    &&op1_entry_0xdd,
+    &&op1_entry_0xde,
+    &&op1_entry_0xdf,
+    &&op1_entry_0xe0,
+    &&op1_entry_0xe1,
+    &&op1_entry_0xe2,
+    &&op1_entry_0xe3,
+    &&op1_entry_0xe4,
+    &&op1_entry_0xe5,
+    &&op1_entry_0xe6,
+    &&op1_entry_0xe7,
+    &&op1_entry_0xe8,
+    &&op1_entry_0xe9,
+    &&op1_entry_0xea,
+    &&op1_entry_0xeb,
+    &&op1_entry_0xec,
+    &&op1_entry_0xed,
+    &&op1_entry_0xee,
+    &&op1_entry_0xef,
+    &&op1_entry_0xf0,
+    &&op1_entry_0xf1,
+    &&op1_entry_0xf2,
+    &&op1_entry_0xf3,
+    &&op1_entry_0xf4,
+    &&op1_entry_0xf5,
+    &&op1_entry_0xf6,
+    &&op1_entry_0xf7,
+    &&op1_entry_0xf8,
+    &&op1_entry_0xf9,
+    &&op1_entry_0xfa,
+    &&op1_entry_0xfb,
+    &&op1_entry_0xfc,
+    &&op1_entry_0xfd,
+    &&op1_entry_0xfe,
+    &&op1_entry_0xff,
+  };
+
+  z80.opcode = RD_BYTE_MEM(z80.reg.pc);
+
+  EXECUTE()
+
+  op0_entry_0x00:
+    op0_0x00();
+    EXECUTE_END();
+
+  op0_entry_0x01:
+    op0_0x01();
+    EXECUTE_END();
+
+  op0_entry_0x02:
+    op0_0x02();
+    EXECUTE_END();
+
+  op0_entry_0x03:
+    op0_0x03();
+    EXECUTE_END();
+
+  op0_entry_0x04:
+    op0_0x04();
+    EXECUTE_END();
+
+  op0_entry_0x05:
+    op0_0x05();
+    EXECUTE_END();
+
+  op0_entry_0x06:
+    op0_0x06();
+    EXECUTE_END();
+
+  op0_entry_0x07:
+    op0_0x07();
+    EXECUTE_END();
+
+  op0_entry_0x08:
+    op0_0x08();
+    EXECUTE_END();
+
+  op0_entry_0x09:
+    op0_0x09();
+    EXECUTE_END();
+
+  op0_entry_0x0a:
+    op0_0x0a();
+    EXECUTE_END();
+
+  op0_entry_0x0b:
+    op0_0x0b();
+    EXECUTE_END();
+
+  op0_entry_0x0c:
+    op0_0x0c();
+    EXECUTE_END();
+
+  op0_entry_0x0d:
+    op0_0x0d();
+    EXECUTE_END();
+
+  op0_entry_0x0e:
+    op0_0x0e();
+    EXECUTE_END();
+
+  op0_entry_0x0f:
+    op0_0x0f();
+    EXECUTE_END();
+
+  op0_entry_0x10:
+    op0_0x10();
+    EXECUTE_END();
+
+  op0_entry_0x11:
+    op0_0x11();
+    EXECUTE_END();
+
+  op0_entry_0x12:
+    op0_0x12();
+    EXECUTE_END();
+
+  op0_entry_0x13:
+    op0_0x13();
+    EXECUTE_END();
+
+  op0_entry_0x14:
+    op0_0x14();
+    EXECUTE_END();
+
+  op0_entry_0x15:
+    op0_0x15();
+    EXECUTE_END();
+
+  op0_entry_0x16:
+    op0_0x16();
+    EXECUTE_END();
+
+  op0_entry_0x17:
+    op0_0x17();
+    EXECUTE_END();
+
+  op0_entry_0x18:
+    op0_0x18();
+    EXECUTE_END();
+
+  op0_entry_0x19:
+    op0_0x19();
+    EXECUTE_END();
+
+  op0_entry_0x1a:
+    op0_0x1a();
+    EXECUTE_END();
+
+  op0_entry_0x1b:
+    op0_0x1b();
+    EXECUTE_END();
+
+  op0_entry_0x1c:
+    op0_0x1c();
+    EXECUTE_END();
+
+  op0_entry_0x1d:
+    op0_0x1d();
+    EXECUTE_END();
+
+  op0_entry_0x1e:
+    op0_0x1e();
+    EXECUTE_END();
+
+  op0_entry_0x1f:
+    op0_0x1f();
+    EXECUTE_END();
+
+  op0_entry_0x20:
+    op0_0x20();
+    EXECUTE_END();
+
+  op0_entry_0x21:
+    op0_0x21();
+    EXECUTE_END();
+
+  op0_entry_0x22:
+    op0_0x22();
+    EXECUTE_END();
+
+  op0_entry_0x23:
+    op0_0x23();
+    EXECUTE_END();
+
+  op0_entry_0x24:
+    op0_0x24();
+    EXECUTE_END();
+
+  op0_entry_0x25:
+    op0_0x25();
+    EXECUTE_END();
+
+  op0_entry_0x26:
+    op0_0x26();
+    EXECUTE_END();
+
+  op0_entry_0x27:
+    op0_0x27();
+    EXECUTE_END();
+
+  op0_entry_0x28:
+    op0_0x28();
+    EXECUTE_END();
+
+  op0_entry_0x29:
+    op0_0x29();
+    EXECUTE_END();
+
+  op0_entry_0x2a:
+    op0_0x2a();
+    EXECUTE_END();
+
+  op0_entry_0x2b:
+    op0_0x2b();
+    EXECUTE_END();
+
+  op0_entry_0x2c:
+    op0_0x2c();
+    EXECUTE_END();
+
+  op0_entry_0x2d:
+    op0_0x2d();
+    EXECUTE_END();
+
+  op0_entry_0x2e:
+    op0_0x2e();
+    EXECUTE_END();
+
+  op0_entry_0x2f:
+    op0_0x2f();
+    EXECUTE_END();
+
+  op0_entry_0x30:
+    op0_0x30();
+    EXECUTE_END();
+
+  op0_entry_0x31:
+    op0_0x31();
+    EXECUTE_END();
+
+  op0_entry_0x32:
+    op0_0x32();
+    EXECUTE_END();
+
+  op0_entry_0x33:
+    op0_0x33();
+    EXECUTE_END();
+
+  op0_entry_0x34:
+    op0_0x34();
+    EXECUTE_END();
+
+  op0_entry_0x35:
+    op0_0x35();
+    EXECUTE_END();
+
+  op0_entry_0x36:
+    op0_0x36();
+    EXECUTE_END();
+
+  op0_entry_0x37:
+    op0_0x37();
+    EXECUTE_END();
+
+  op0_entry_0x38:
+    op0_0x38();
+    EXECUTE_END();
+
+  op0_entry_0x39:
+    op0_0x39();
+    EXECUTE_END();
+
+  op0_entry_0x3a:
+    op0_0x3a();
+    EXECUTE_END();
+
+  op0_entry_0x3b:
+    op0_0x3b();
+    EXECUTE_END();
+
+  op0_entry_0x3c:
+    op0_0x3c();
+    EXECUTE_END();
+
+  op0_entry_0x3d:
+    op0_0x3d();
+    EXECUTE_END();
+
+  op0_entry_0x3e:
+    op0_0x3e();
+    EXECUTE_END();
+
+  op0_entry_0x3f:
+    op0_0x3f();
+    EXECUTE_END();
+
+  op0_entry_0x40:
+    op0_0x40();
+    EXECUTE_END();
+
+  op0_entry_0x41:
+    op0_0x41();
+    EXECUTE_END();
+
+  op0_entry_0x42:
+    op0_0x42();
+    EXECUTE_END();
+
+  op0_entry_0x43:
+    op0_0x43();
+    EXECUTE_END();
+
+  op0_entry_0x44:
+    op0_0x44();
+    EXECUTE_END();
+
+  op0_entry_0x45:
+    op0_0x45();
+    EXECUTE_END();
+
+  op0_entry_0x46:
+    op0_0x46();
+    EXECUTE_END();
+
+  op0_entry_0x47:
+    op0_0x47();
+    EXECUTE_END();
+
+  op0_entry_0x48:
+    op0_0x48();
+    EXECUTE_END();
+
+  op0_entry_0x49:
+    op0_0x49();
+    EXECUTE_END();
+
+  op0_entry_0x4a:
+    op0_0x4a();
+    EXECUTE_END();
+
+  op0_entry_0x4b:
+    op0_0x4b();
+    EXECUTE_END();
+
+  op0_entry_0x4c:
+    op0_0x4c();
+    EXECUTE_END();
+
+  op0_entry_0x4d:
+    op0_0x4d();
+    EXECUTE_END();
+
+  op0_entry_0x4e:
+    op0_0x4e();
+    EXECUTE_END();
+
+  op0_entry_0x4f:
+    op0_0x4f();
+    EXECUTE_END();
+
+  op0_entry_0x50:
+    op0_0x50();
+    EXECUTE_END();
+
+  op0_entry_0x51:
+    op0_0x51();
+    EXECUTE_END();
+
+  op0_entry_0x52:
+    op0_0x52();
+    EXECUTE_END();
+
+  op0_entry_0x53:
+    op0_0x53();
+    EXECUTE_END();
+
+  op0_entry_0x54:
+    op0_0x54();
+    EXECUTE_END();
+
+  op0_entry_0x55:
+    op0_0x55();
+    EXECUTE_END();
+
+  op0_entry_0x56:
+    op0_0x56();
+    EXECUTE_END();
+
+  op0_entry_0x57:
+    op0_0x57();
+    EXECUTE_END();
+
+  op0_entry_0x58:
+    op0_0x58();
+    EXECUTE_END();
+
+  op0_entry_0x59:
+    op0_0x59();
+    EXECUTE_END();
+
+  op0_entry_0x5a:
+    op0_0x5a();
+    EXECUTE_END();
+
+  op0_entry_0x5b:
+    op0_0x5b();
+    EXECUTE_END();
+
+  op0_entry_0x5c:
+    op0_0x5c();
+    EXECUTE_END();
+
+  op0_entry_0x5d:
+    op0_0x5d();
+    EXECUTE_END();
+
+  op0_entry_0x5e:
+    op0_0x5e();
+    EXECUTE_END();
+
+  op0_entry_0x5f:
+    op0_0x5f();
+    EXECUTE_END();
+
+  op0_entry_0x60:
+    op0_0x60();
+    EXECUTE_END();
+
+  op0_entry_0x61:
+    op0_0x61();
+    EXECUTE_END();
+
+  op0_entry_0x62:
+    op0_0x62();
+    EXECUTE_END();
+
+  op0_entry_0x63:
+    op0_0x63();
+    EXECUTE_END();
+
+  op0_entry_0x64:
+    op0_0x64();
+    EXECUTE_END();
+
+  op0_entry_0x65:
+    op0_0x65();
+    EXECUTE_END();
+
+  op0_entry_0x66:
+    op0_0x66();
+    EXECUTE_END();
+
+  op0_entry_0x67:
+    op0_0x67();
+    EXECUTE_END();
+
+  op0_entry_0x68:
+    op0_0x68();
+    EXECUTE_END();
+
+  op0_entry_0x69:
+    op0_0x69();
+    EXECUTE_END();
+
+  op0_entry_0x6a:
+    op0_0x6a();
+    EXECUTE_END();
+
+  op0_entry_0x6b:
+    op0_0x6b();
+    EXECUTE_END();
+
+  op0_entry_0x6c:
+    op0_0x6c();
+    EXECUTE_END();
+
+  op0_entry_0x6d:
+    op0_0x6d();
+    EXECUTE_END();
+
+  op0_entry_0x6e:
+    op0_0x6e();
+    EXECUTE_END();
+
+  op0_entry_0x6f:
+    op0_0x6f();
+    EXECUTE_END();
+
+  op0_entry_0x70:
+    op0_0x70();
+    EXECUTE_END();
+
+  op0_entry_0x71:
+    op0_0x71();
+    EXECUTE_END();
+
+  op0_entry_0x72:
+    op0_0x72();
+    EXECUTE_END();
+
+  op0_entry_0x73:
+    op0_0x73();
+    EXECUTE_END();
+
+  op0_entry_0x74:
+    op0_0x74();
+    EXECUTE_END();
+
+  op0_entry_0x75:
+    op0_0x75();
+    EXECUTE_END();
+
+  op0_entry_0x76:
+    op0_0x76();
+    EXECUTE_END();
+
+  op0_entry_0x77:
+    op0_0x77();
+    EXECUTE_END();
+
+  op0_entry_0x78:
+    op0_0x78();
+    EXECUTE_END();
+
+  op0_entry_0x79:
+    op0_0x79();
+    EXECUTE_END();
+
+  op0_entry_0x7a:
+    op0_0x7a();
+    EXECUTE_END();
+
+  op0_entry_0x7b:
+    op0_0x7b();
+    EXECUTE_END();
+
+  op0_entry_0x7c:
+    op0_0x7c();
+    EXECUTE_END();
+
+  op0_entry_0x7d:
+    op0_0x7d();
+    EXECUTE_END();
+
+  op0_entry_0x7e:
+    op0_0x7e();
+    EXECUTE_END();
+
+  op0_entry_0x7f:
+    op0_0x7f();
+    EXECUTE_END();
+
+  op0_entry_0x80:
+    op0_0x80();
+    EXECUTE_END();
+
+  op0_entry_0x81:
+    op0_0x81();
+    EXECUTE_END();
+
+  op0_entry_0x82:
+    op0_0x82();
+    EXECUTE_END();
+
+  op0_entry_0x83:
+    op0_0x83();
+    EXECUTE_END();
+
+  op0_entry_0x84:
+    op0_0x84();
+    EXECUTE_END();
+
+  op0_entry_0x85:
+    op0_0x85();
+    EXECUTE_END();
+
+  op0_entry_0x86:
+    op0_0x86();
+    EXECUTE_END();
+
+  op0_entry_0x87:
+    op0_0x87();
+    EXECUTE_END();
+
+  op0_entry_0x88:
+    op0_0x88();
+    EXECUTE_END();
+
+  op0_entry_0x89:
+    op0_0x89();
+    EXECUTE_END();
+
+  op0_entry_0x8a:
+    op0_0x8a();
+    EXECUTE_END();
+
+  op0_entry_0x8b:
+    op0_0x8b();
+    EXECUTE_END();
+
+  op0_entry_0x8c:
+    op0_0x8c();
+    EXECUTE_END();
+
+  op0_entry_0x8d:
+    op0_0x8d();
+    EXECUTE_END();
+
+  op0_entry_0x8e:
+    op0_0x8e();
+    EXECUTE_END();
+
+  op0_entry_0x8f:
+    op0_0x8f();
+    EXECUTE_END();
+
+  op0_entry_0x90:
+    op0_0x90();
+    EXECUTE_END();
+
+  op0_entry_0x91:
+    op0_0x91();
+    EXECUTE_END();
+
+  op0_entry_0x92:
+    op0_0x92();
+    EXECUTE_END();
+
+  op0_entry_0x93:
+    op0_0x93();
+    EXECUTE_END();
+
+  op0_entry_0x94:
+    op0_0x94();
+    EXECUTE_END();
+
+  op0_entry_0x95:
+    op0_0x95();
+    EXECUTE_END();
+
+  op0_entry_0x96:
+    op0_0x96();
+    EXECUTE_END();
+
+  op0_entry_0x97:
+    op0_0x97();
+    EXECUTE_END();
+
+  op0_entry_0x98:
+    op0_0x98();
+    EXECUTE_END();
+
+  op0_entry_0x99:
+    op0_0x99();
+    EXECUTE_END();
+
+  op0_entry_0x9a:
+    op0_0x9a();
+    EXECUTE_END();
+
+  op0_entry_0x9b:
+    op0_0x9b();
+    EXECUTE_END();
+
+  op0_entry_0x9c:
+    op0_0x9c();
+    EXECUTE_END();
+
+  op0_entry_0x9d:
+    op0_0x9d();
+    EXECUTE_END();
+
+  op0_entry_0x9e:
+    op0_0x9e();
+    EXECUTE_END();
+
+  op0_entry_0x9f:
+    op0_0x9f();
+    EXECUTE_END();
+
+  op0_entry_0xa0:
+    op0_0xa0();
+    EXECUTE_END();
+
+  op0_entry_0xa1:
+    op0_0xa1();
+    EXECUTE_END();
+
+  op0_entry_0xa2:
+    op0_0xa2();
+    EXECUTE_END();
+
+  op0_entry_0xa3:
+    op0_0xa3();
+    EXECUTE_END();
+
+  op0_entry_0xa4:
+    op0_0xa4();
+    EXECUTE_END();
+
+  op0_entry_0xa5:
+    op0_0xa5();
+    EXECUTE_END();
+
+  op0_entry_0xa6:
+    op0_0xa6();
+    EXECUTE_END();
+
+  op0_entry_0xa7:
+    op0_0xa7();
+    EXECUTE_END();
+
+  op0_entry_0xa8:
+    op0_0xa8();
+    EXECUTE_END();
+
+  op0_entry_0xa9:
+    op0_0xa9();
+    EXECUTE_END();
+
+  op0_entry_0xaa:
+    op0_0xaa();
+    EXECUTE_END();
+
+  op0_entry_0xab:
+    op0_0xab();
+    EXECUTE_END();
+
+  op0_entry_0xac:
+    op0_0xac();
+    EXECUTE_END();
+
+  op0_entry_0xad:
+    op0_0xad();
+    EXECUTE_END();
+
+  op0_entry_0xae:
+    op0_0xae();
+    EXECUTE_END();
+
+  op0_entry_0xaf:
+    op0_0xaf();
+    EXECUTE_END();
+
+  op0_entry_0xb0:
+    op0_0xb0();
+    EXECUTE_END();
+
+  op0_entry_0xb1:
+    op0_0xb1();
+    EXECUTE_END();
+
+  op0_entry_0xb2:
+    op0_0xb2();
+    EXECUTE_END();
+
+  op0_entry_0xb3:
+    op0_0xb3();
+    EXECUTE_END();
+
+  op0_entry_0xb4:
+    op0_0xb4();
+    EXECUTE_END();
+
+  op0_entry_0xb5:
+    op0_0xb5();
+    EXECUTE_END();
+
+  op0_entry_0xb6:
+    op0_0xb6();
+    EXECUTE_END();
+
+  op0_entry_0xb7:
+    op0_0xb7();
+    EXECUTE_END();
+
+  op0_entry_0xb8:
+    op0_0xb8();
+    EXECUTE_END();
+
+  op0_entry_0xb9:
+    op0_0xb9();
+    EXECUTE_END();
+
+  op0_entry_0xba:
+    op0_0xba();
+    EXECUTE_END();
+
+  op0_entry_0xbb:
+    op0_0xbb();
+    EXECUTE_END();
+
+  op0_entry_0xbc:
+    op0_0xbc();
+    EXECUTE_END();
+
+  op0_entry_0xbd:
+    op0_0xbd();
+    EXECUTE_END();
+
+  op0_entry_0xbe:
+    op0_0xbe();
+    EXECUTE_END();
+
+  op0_entry_0xbf:
+    op0_0xbf();
+    EXECUTE_END();
+
+  op0_entry_0xc0:
+    op0_0xc0();
+    EXECUTE_END();
+
+  op0_entry_0xc1:
+    op0_0xc1();
+    EXECUTE_END();
+
+  op0_entry_0xc2:
+    op0_0xc2();
+    EXECUTE_END();
+
+  op0_entry_0xc3:
+    op0_0xc3();
+    EXECUTE_END();
+
+  op0_entry_0xc4:
+    op0_0xc4();
+    EXECUTE_END();
+
+  op0_entry_0xc5:
+    op0_0xc5();
+    EXECUTE_END();
+
+  op0_entry_0xc6:
+    op0_0xc6();
+    EXECUTE_END();
+
+  op0_entry_0xc7:
+    op0_0xc7();
+    EXECUTE_END();
+
+  op0_entry_0xc8:
+    op0_0xc8();
+    EXECUTE_END();
+
+  op0_entry_0xc9:
+    op0_0xc9();
+    EXECUTE_END();
+
+  op0_entry_0xca:
+    op0_0xca();
+    EXECUTE_END();
+
+  op0_entry_0xcb:
+    z80.opcode = RD_BYTE_MEM(z80.reg.pc+1);
+    EXECUTE_CB();
+
+  op0_entry_0xcc:
+    op0_0xcc();
+    EXECUTE_END();
+
+  op0_entry_0xcd:
+    op0_0xcd();
+    EXECUTE_END();
+
+  op0_entry_0xce:
+    op0_0xce();
+    EXECUTE_END();
+
+  op0_entry_0xcf:
+    op0_0xcf();
+    EXECUTE_END();
+
+  op0_entry_0xd0:
+    op0_0xd0();
+    EXECUTE_END();
+
+  op0_entry_0xd1:
+    op0_0xd1();
+    EXECUTE_END();
+
+  op0_entry_0xd2:
+    op0_0xd2();
+    EXECUTE_END();
+
+  op0_entry_0xd3:
+    op0_0xd3();
+    EXECUTE_END();
+
+  op0_entry_0xd4:
+    op0_0xd4();
+    EXECUTE_END();
+
+  op0_entry_0xd5:
+    op0_0xd5();
+    EXECUTE_END();
+
+  op0_entry_0xd6:
+    op0_0xd6();
+    EXECUTE_END();
+
+  op0_entry_0xd7:
+    op0_0xd7();
+    EXECUTE_END();
+
+  op0_entry_0xd8:
+    op0_0xd8();
+    EXECUTE_END();
+
+  op0_entry_0xd9:
+    op0_0xd9();
+    EXECUTE_END();
+
+  op0_entry_0xda:
+    op0_0xda();
+    EXECUTE_END();
+
+  op0_entry_0xdb:
+    op0_0xdb();
+    EXECUTE_END();
+
+  op0_entry_0xdc:
+    op0_0xdc();
+    EXECUTE_END();
+
+  op0_entry_0xdd:
+    op0_0xdd();
+    EXECUTE_END();
+
+  op0_entry_0xde:
+    op0_0xde();
+    EXECUTE_END();
+
+  op0_entry_0xdf:
+    op0_0xdf();
+    EXECUTE_END();
+
+  op0_entry_0xe0:
+    op0_0xe0();
+    EXECUTE_END();
+
+  op0_entry_0xe1:
+    op0_0xe1();
+    EXECUTE_END();
+
+  op0_entry_0xe2:
+    op0_0xe2();
+    EXECUTE_END();
+
+  op0_entry_0xe3:
+    op0_0xe3();
+    EXECUTE_END();
+
+  op0_entry_0xe4:
+    op0_0xe4();
+    EXECUTE_END();
+
+  op0_entry_0xe5:
+    op0_0xe5();
+    EXECUTE_END();
+
+  op0_entry_0xe6:
+    op0_0xe6();
+    EXECUTE_END();
+
+  op0_entry_0xe7:
+    op0_0xe7();
+    EXECUTE_END();
+
+  op0_entry_0xe8:
+    op0_0xe8();
+    EXECUTE_END();
+
+  op0_entry_0xe9:
+    op0_0xe9();
+    EXECUTE_END();
+
+  op0_entry_0xea:
+    op0_0xea();
+    EXECUTE_END();
+
+  op0_entry_0xeb:
+    op0_0xeb();
+    EXECUTE_END();
+
+  op0_entry_0xec:
+    op0_0xec();
+    EXECUTE_END();
+
+  op0_entry_0xed:
+    op0_0xed();
+    EXECUTE_END();
+
+  op0_entry_0xee:
+    op0_0xee();
+    EXECUTE_END();
+
+  op0_entry_0xef:
+    op0_0xef();
+    EXECUTE_END();
+
+  op0_entry_0xf0:
+    op0_0xf0();
+    EXECUTE_END();
+
+  op0_entry_0xf1:
+    op0_0xf1();
+    EXECUTE_END();
+
+  op0_entry_0xf2:
+    op0_0xf2();
+    EXECUTE_END();
+
+  op0_entry_0xf3:
+    op0_0xf3();
+    EXECUTE_END();
+
+  op0_entry_0xf4:
+    op0_0xf4();
+    EXECUTE_END();
+
+  op0_entry_0xf5:
+    op0_0xf5();
+    EXECUTE_END();
+
+  op0_entry_0xf6:
+    op0_0xf6();
+    EXECUTE_END();
+
+  op0_entry_0xf7:
+    op0_0xf7();
+    EXECUTE_END();
+
+  op0_entry_0xf8:
+    op0_0xf8();
+    EXECUTE_END();
+
+  op0_entry_0xf9:
+    op0_0xf9();
+    EXECUTE_END();
+
+  op0_entry_0xfa:
+    op0_0xfa();
+    EXECUTE_END();
+
+  op0_entry_0xfb:
+    op0_0xfb();
+    EXECUTE_END();
+
+  op0_entry_0xfc:
+    op0_0xfc();
+    EXECUTE_END();
+
+  op0_entry_0xfd:
+    op0_0xfd();
+    EXECUTE_END();
+
+  op0_entry_0xfe:
+    op0_0xfe();
+    EXECUTE_END();
+
+  op0_entry_0xff:
+    op0_0xff();
+    EXECUTE_END();
+
+  op1_entry_0x00:
+    op1_0x00();
+    EXECUTE_CB_END();
+
+  op1_entry_0x01:
+    op1_0x01();
+    EXECUTE_CB_END();
+
+  op1_entry_0x02:
+    op1_0x02();
+    EXECUTE_CB_END();
+
+  op1_entry_0x03:
+    op1_0x03();
+    EXECUTE_CB_END();
+
+  op1_entry_0x04:
+    op1_0x04();
+    EXECUTE_CB_END();
+
+  op1_entry_0x05:
+    op1_0x05();
+    EXECUTE_CB_END();
+
+  op1_entry_0x06:
+    op1_0x06();
+    EXECUTE_CB_END();
+
+  op1_entry_0x07:
+    op1_0x07();
+    EXECUTE_CB_END();
+
+  op1_entry_0x08:
+    op1_0x08();
+    EXECUTE_CB_END();
+
+  op1_entry_0x09:
+    op1_0x09();
+    EXECUTE_CB_END();
+
+  op1_entry_0x0a:
+    op1_0x0a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x0b:
+    op1_0x0b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x0c:
+    op1_0x0c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x0d:
+    op1_0x0d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x0e:
+    op1_0x0e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x0f:
+    op1_0x0f();
+    EXECUTE_CB_END();
+
+  op1_entry_0x10:
+    op1_0x10();
+    EXECUTE_CB_END();
+
+  op1_entry_0x11:
+    op1_0x11();
+    EXECUTE_CB_END();
+
+  op1_entry_0x12:
+    op1_0x12();
+    EXECUTE_CB_END();
+
+  op1_entry_0x13:
+    op1_0x13();
+    EXECUTE_CB_END();
+
+  op1_entry_0x14:
+    op1_0x14();
+    EXECUTE_CB_END();
+
+  op1_entry_0x15:
+    op1_0x15();
+    EXECUTE_CB_END();
+
+  op1_entry_0x16:
+    op1_0x16();
+    EXECUTE_CB_END();
+
+  op1_entry_0x17:
+    op1_0x17();
+    EXECUTE_CB_END();
+
+  op1_entry_0x18:
+    op1_0x18();
+    EXECUTE_CB_END();
+
+  op1_entry_0x19:
+    op1_0x19();
+    EXECUTE_CB_END();
+
+  op1_entry_0x1a:
+    op1_0x1a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x1b:
+    op1_0x1b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x1c:
+    op1_0x1c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x1d:
+    op1_0x1d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x1e:
+    op1_0x1e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x1f:
+    op1_0x1f();
+    EXECUTE_CB_END();
+
+  op1_entry_0x20:
+    op1_0x20();
+    EXECUTE_CB_END();
+
+  op1_entry_0x21:
+    op1_0x21();
+    EXECUTE_CB_END();
+
+  op1_entry_0x22:
+    op1_0x22();
+    EXECUTE_CB_END();
+
+  op1_entry_0x23:
+    op1_0x23();
+    EXECUTE_CB_END();
+
+  op1_entry_0x24:
+    op1_0x24();
+    EXECUTE_CB_END();
+
+  op1_entry_0x25:
+    op1_0x25();
+    EXECUTE_CB_END();
+
+  op1_entry_0x26:
+    op1_0x26();
+    EXECUTE_CB_END();
+
+  op1_entry_0x27:
+    op1_0x27();
+    EXECUTE_CB_END();
+
+  op1_entry_0x28:
+    op1_0x28();
+    EXECUTE_CB_END();
+
+  op1_entry_0x29:
+    op1_0x29();
+    EXECUTE_CB_END();
+
+  op1_entry_0x2a:
+    op1_0x2a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x2b:
+    op1_0x2b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x2c:
+    op1_0x2c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x2d:
+    op1_0x2d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x2e:
+    op1_0x2e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x2f:
+    op1_0x2f();
+    EXECUTE_CB_END();
+
+  op1_entry_0x30:
+    op1_0x30();
+    EXECUTE_CB_END();
+
+  op1_entry_0x31:
+    op1_0x31();
+    EXECUTE_CB_END();
+
+  op1_entry_0x32:
+    op1_0x32();
+    EXECUTE_CB_END();
+
+  op1_entry_0x33:
+    op1_0x33();
+    EXECUTE_CB_END();
+
+  op1_entry_0x34:
+    op1_0x34();
+    EXECUTE_CB_END();
+
+  op1_entry_0x35:
+    op1_0x35();
+    EXECUTE_CB_END();
+
+  op1_entry_0x36:
+    op1_0x36();
+    EXECUTE_CB_END();
+
+  op1_entry_0x37:
+    op1_0x37();
+    EXECUTE_CB_END();
+
+  op1_entry_0x38:
+    op1_0x38();
+    EXECUTE_CB_END();
+
+  op1_entry_0x39:
+    op1_0x39();
+    EXECUTE_CB_END();
+
+  op1_entry_0x3a:
+    op1_0x3a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x3b:
+    op1_0x3b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x3c:
+    op1_0x3c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x3d:
+    op1_0x3d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x3e:
+    op1_0x3e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x3f:
+    op1_0x3f();
+    EXECUTE_CB_END();
+
+  op1_entry_0x40:
+    op1_0x40();
+    EXECUTE_CB_END();
+
+  op1_entry_0x41:
+    op1_0x41();
+    EXECUTE_CB_END();
+
+  op1_entry_0x42:
+    op1_0x42();
+    EXECUTE_CB_END();
+
+  op1_entry_0x43:
+    op1_0x43();
+    EXECUTE_CB_END();
+
+  op1_entry_0x44:
+    op1_0x44();
+    EXECUTE_CB_END();
+
+  op1_entry_0x45:
+    op1_0x45();
+    EXECUTE_CB_END();
+
+  op1_entry_0x46:
+    op1_0x46();
+    EXECUTE_CB_END();
+
+  op1_entry_0x47:
+    op1_0x47();
+    EXECUTE_CB_END();
+
+  op1_entry_0x48:
+    op1_0x48();
+    EXECUTE_CB_END();
+
+  op1_entry_0x49:
+    op1_0x49();
+    EXECUTE_CB_END();
+
+  op1_entry_0x4a:
+    op1_0x4a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x4b:
+    op1_0x4b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x4c:
+    op1_0x4c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x4d:
+    op1_0x4d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x4e:
+    op1_0x4e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x4f:
+    op1_0x4f();
+    EXECUTE_CB_END();
+
+  op1_entry_0x50:
+    op1_0x50();
+    EXECUTE_CB_END();
+
+  op1_entry_0x51:
+    op1_0x51();
+    EXECUTE_CB_END();
+
+  op1_entry_0x52:
+    op1_0x52();
+    EXECUTE_CB_END();
+
+  op1_entry_0x53:
+    op1_0x53();
+    EXECUTE_CB_END();
+
+  op1_entry_0x54:
+    op1_0x54();
+    EXECUTE_CB_END();
+
+  op1_entry_0x55:
+    op1_0x55();
+    EXECUTE_CB_END();
+
+  op1_entry_0x56:
+    op1_0x56();
+    EXECUTE_CB_END();
+
+  op1_entry_0x57:
+    op1_0x57();
+    EXECUTE_CB_END();
+
+  op1_entry_0x58:
+    op1_0x58();
+    EXECUTE_CB_END();
+
+  op1_entry_0x59:
+    op1_0x59();
+    EXECUTE_CB_END();
+
+  op1_entry_0x5a:
+    op1_0x5a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x5b:
+    op1_0x5b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x5c:
+    op1_0x5c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x5d:
+    op1_0x5d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x5e:
+    op1_0x5e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x5f:
+    op1_0x5f();
+    EXECUTE_CB_END();
+
+  op1_entry_0x60:
+    op1_0x60();
+    EXECUTE_CB_END();
+
+  op1_entry_0x61:
+    op1_0x61();
+    EXECUTE_CB_END();
+
+  op1_entry_0x62:
+    op1_0x62();
+    EXECUTE_CB_END();
+
+  op1_entry_0x63:
+    op1_0x63();
+    EXECUTE_CB_END();
+
+  op1_entry_0x64:
+    op1_0x64();
+    EXECUTE_CB_END();
+
+  op1_entry_0x65:
+    op1_0x65();
+    EXECUTE_CB_END();
+
+  op1_entry_0x66:
+    op1_0x66();
+    EXECUTE_CB_END();
+
+  op1_entry_0x67:
+    op1_0x67();
+    EXECUTE_CB_END();
+
+  op1_entry_0x68:
+    op1_0x68();
+    EXECUTE_CB_END();
+
+  op1_entry_0x69:
+    op1_0x69();
+    EXECUTE_CB_END();
+
+  op1_entry_0x6a:
+    op1_0x6a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x6b:
+    op1_0x6b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x6c:
+    op1_0x6c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x6d:
+    op1_0x6d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x6e:
+    op1_0x6e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x6f:
+    op1_0x6f();
+    EXECUTE_CB_END();
+
+  op1_entry_0x70:
+    op1_0x70();
+    EXECUTE_CB_END();
+
+  op1_entry_0x71:
+    op1_0x71();
+    EXECUTE_CB_END();
+
+  op1_entry_0x72:
+    op1_0x72();
+    EXECUTE_CB_END();
+
+  op1_entry_0x73:
+    op1_0x73();
+    EXECUTE_CB_END();
+
+  op1_entry_0x74:
+    op1_0x74();
+    EXECUTE_CB_END();
+
+  op1_entry_0x75:
+    op1_0x75();
+    EXECUTE_CB_END();
+
+  op1_entry_0x76:
+    op1_0x76();
+    EXECUTE_CB_END();
+
+  op1_entry_0x77:
+    op1_0x77();
+    EXECUTE_CB_END();
+
+  op1_entry_0x78:
+    op1_0x78();
+    EXECUTE_CB_END();
+
+  op1_entry_0x79:
+    op1_0x79();
+    EXECUTE_CB_END();
+
+  op1_entry_0x7a:
+    op1_0x7a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x7b:
+    op1_0x7b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x7c:
+    op1_0x7c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x7d:
+    op1_0x7d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x7e:
+    op1_0x7e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x7f:
+    op1_0x7f();
+    EXECUTE_CB_END();
+
+  op1_entry_0x80:
+    op1_0x80();
+    EXECUTE_CB_END();
+
+  op1_entry_0x81:
+    op1_0x81();
+    EXECUTE_CB_END();
+
+  op1_entry_0x82:
+    op1_0x82();
+    EXECUTE_CB_END();
+
+  op1_entry_0x83:
+    op1_0x83();
+    EXECUTE_CB_END();
+
+  op1_entry_0x84:
+    op1_0x84();
+    EXECUTE_CB_END();
+
+  op1_entry_0x85:
+    op1_0x85();
+    EXECUTE_CB_END();
+
+  op1_entry_0x86:
+    op1_0x86();
+    EXECUTE_CB_END();
+
+  op1_entry_0x87:
+    op1_0x87();
+    EXECUTE_CB_END();
+
+  op1_entry_0x88:
+    op1_0x88();
+    EXECUTE_CB_END();
+
+  op1_entry_0x89:
+    op1_0x89();
+    EXECUTE_CB_END();
+
+  op1_entry_0x8a:
+    op1_0x8a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x8b:
+    op1_0x8b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x8c:
+    op1_0x8c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x8d:
+    op1_0x8d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x8e:
+    op1_0x8e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x8f:
+    op1_0x8f();
+    EXECUTE_CB_END();
+
+  op1_entry_0x90:
+    op1_0x90();
+    EXECUTE_CB_END();
+
+  op1_entry_0x91:
+    op1_0x91();
+    EXECUTE_CB_END();
+
+  op1_entry_0x92:
+    op1_0x92();
+    EXECUTE_CB_END();
+
+  op1_entry_0x93:
+    op1_0x93();
+    EXECUTE_CB_END();
+
+  op1_entry_0x94:
+    op1_0x94();
+    EXECUTE_CB_END();
+
+  op1_entry_0x95:
+    op1_0x95();
+    EXECUTE_CB_END();
+
+  op1_entry_0x96:
+    op1_0x96();
+    EXECUTE_CB_END();
+
+  op1_entry_0x97:
+    op1_0x97();
+    EXECUTE_CB_END();
+
+  op1_entry_0x98:
+    op1_0x98();
+    EXECUTE_CB_END();
+
+  op1_entry_0x99:
+    op1_0x99();
+    EXECUTE_CB_END();
+
+  op1_entry_0x9a:
+    op1_0x9a();
+    EXECUTE_CB_END();
+
+  op1_entry_0x9b:
+    op1_0x9b();
+    EXECUTE_CB_END();
+
+  op1_entry_0x9c:
+    op1_0x9c();
+    EXECUTE_CB_END();
+
+  op1_entry_0x9d:
+    op1_0x9d();
+    EXECUTE_CB_END();
+
+  op1_entry_0x9e:
+    op1_0x9e();
+    EXECUTE_CB_END();
+
+  op1_entry_0x9f:
+    op1_0x9f();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa0:
+    op1_0xa0();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa1:
+    op1_0xa1();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa2:
+    op1_0xa2();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa3:
+    op1_0xa3();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa4:
+    op1_0xa4();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa5:
+    op1_0xa5();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa6:
+    op1_0xa6();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa7:
+    op1_0xa7();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa8:
+    op1_0xa8();
+    EXECUTE_CB_END();
+
+  op1_entry_0xa9:
+    op1_0xa9();
+    EXECUTE_CB_END();
+
+  op1_entry_0xaa:
+    op1_0xaa();
+    EXECUTE_CB_END();
+
+  op1_entry_0xab:
+    op1_0xab();
+    EXECUTE_CB_END();
+
+  op1_entry_0xac:
+    op1_0xac();
+    EXECUTE_CB_END();
+
+  op1_entry_0xad:
+    op1_0xad();
+    EXECUTE_CB_END();
+
+  op1_entry_0xae:
+    op1_0xae();
+    EXECUTE_CB_END();
+
+  op1_entry_0xaf:
+    op1_0xaf();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb0:
+    op1_0xb0();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb1:
+    op1_0xb1();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb2:
+    op1_0xb2();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb3:
+    op1_0xb3();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb4:
+    op1_0xb4();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb5:
+    op1_0xb5();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb6:
+    op1_0xb6();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb7:
+    op1_0xb7();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb8:
+    op1_0xb8();
+    EXECUTE_CB_END();
+
+  op1_entry_0xb9:
+    op1_0xb9();
+    EXECUTE_CB_END();
+
+  op1_entry_0xba:
+    op1_0xba();
+    EXECUTE_CB_END();
+
+  op1_entry_0xbb:
+    op1_0xbb();
+    EXECUTE_CB_END();
+
+  op1_entry_0xbc:
+    op1_0xbc();
+    EXECUTE_CB_END();
+
+  op1_entry_0xbd:
+    op1_0xbd();
+    EXECUTE_CB_END();
+
+  op1_entry_0xbe:
+    op1_0xbe();
+    EXECUTE_CB_END();
+
+  op1_entry_0xbf:
+    op1_0xbf();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc0:
+    op1_0xc0();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc1:
+    op1_0xc1();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc2:
+    op1_0xc2();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc3:
+    op1_0xc3();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc4:
+    op1_0xc4();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc5:
+    op1_0xc5();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc6:
+    op1_0xc6();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc7:
+    op1_0xc7();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc8:
+    op1_0xc8();
+    EXECUTE_CB_END();
+
+  op1_entry_0xc9:
+    op1_0xc9();
+    EXECUTE_CB_END();
+
+  op1_entry_0xca:
+    op1_0xca();
+    EXECUTE_CB_END();
+
+  op1_entry_0xcb:
+    op1_0xcb();
+    EXECUTE_CB_END();
+
+  op1_entry_0xcc:
+    op1_0xcc();
+    EXECUTE_CB_END();
+
+  op1_entry_0xcd:
+    op1_0xcd();
+    EXECUTE_CB_END();
+
+  op1_entry_0xce:
+    op1_0xce();
+    EXECUTE_CB_END();
+
+  op1_entry_0xcf:
+    op1_0xcf();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd0:
+    op1_0xd0();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd1:
+    op1_0xd1();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd2:
+    op1_0xd2();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd3:
+    op1_0xd3();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd4:
+    op1_0xd4();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd5:
+    op1_0xd5();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd6:
+    op1_0xd6();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd7:
+    op1_0xd7();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd8:
+    op1_0xd8();
+    EXECUTE_CB_END();
+
+  op1_entry_0xd9:
+    op1_0xd9();
+    EXECUTE_CB_END();
+
+  op1_entry_0xda:
+    op1_0xda();
+    EXECUTE_CB_END();
+
+  op1_entry_0xdb:
+    op1_0xdb();
+    EXECUTE_CB_END();
+
+  op1_entry_0xdc:
+    op1_0xdc();
+    EXECUTE_CB_END();
+
+  op1_entry_0xdd:
+    op1_0xdd();
+    EXECUTE_CB_END();
+
+  op1_entry_0xde:
+    op1_0xde();
+    EXECUTE_CB_END();
+
+  op1_entry_0xdf:
+    op1_0xdf();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe0:
+    op1_0xe0();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe1:
+    op1_0xe1();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe2:
+    op1_0xe2();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe3:
+    op1_0xe3();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe4:
+    op1_0xe4();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe5:
+    op1_0xe5();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe6:
+    op1_0xe6();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe7:
+    op1_0xe7();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe8:
+    op1_0xe8();
+    EXECUTE_CB_END();
+
+  op1_entry_0xe9:
+    op1_0xe9();
+    EXECUTE_CB_END();
+
+  op1_entry_0xea:
+    op1_0xea();
+    EXECUTE_CB_END();
+
+  op1_entry_0xeb:
+    op1_0xeb();
+    EXECUTE_CB_END();
+
+  op1_entry_0xec:
+    op1_0xec();
+    EXECUTE_CB_END();
+
+  op1_entry_0xed:
+    op1_0xed();
+    EXECUTE_CB_END();
+
+  op1_entry_0xee:
+    op1_0xee();
+    EXECUTE_CB_END();
+
+  op1_entry_0xef:
+    op1_0xef();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf0:
+    op1_0xf0();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf1:
+    op1_0xf1();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf2:
+    op1_0xf2();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf3:
+    op1_0xf3();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf4:
+    op1_0xf4();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf5:
+    op1_0xf5();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf6:
+    op1_0xf6();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf7:
+    op1_0xf7();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf8:
+    op1_0xf8();
+    EXECUTE_CB_END();
+
+  op1_entry_0xf9:
+    op1_0xf9();
+    EXECUTE_CB_END();
+
+  op1_entry_0xfa:
+    op1_0xfa();
+    EXECUTE_CB_END();
+
+  op1_entry_0xfb:
+    op1_0xfb();
+    EXECUTE_CB_END();
+
+  op1_entry_0xfc:
+    op1_0xfc();
+    EXECUTE_CB_END();
+
+  op1_entry_0xfd:
+    op1_0xfd();
+    EXECUTE_CB_END();
+
+  op1_entry_0xfe:
+    op1_0xfe();
+    EXECUTE_CB_END();
+
+  op1_entry_0xff:
+    op1_0xff();
+    EXECUTE_CB_END();
+
+  single_step_end:
+    z80.cycles = cycles[z80.opcode];
+    return;
+
+  single_step_cb_end:
+    z80.cycles = cycles_cb[z80.opcode];
+    return;
+}
+
+#endif
+#endif
